Project1/GridMap.cpp

```diff
@@ -0,0 +1,148 @@
+
+//Constructors//
+/////////////////////////////////////////////////////////////////////////
+GridMap::GridMap(int width, int height) :
+	Grid(width, height)
+{}
+
+GridMap::GridMap(int width, int height, std::vector<PartyData> data) :
+	Grid(width, height, data)
+{}
+
+//Party Access//
+/////////////////////////////////////////////////////////////////////////
+Party& GridMap::getParty(const Unit & unit) {
+	for (auto & party : _parties) {
+		if (*unit.getParty() == party) {
+			return party;
+		}
+	}
+	Expects(false);
+}
+
+const Party& GridMap::getParty(const Unit & unit) const {
+	return unit.getParty()->getParty(Passkey<GridMap>());
+}
+
+//Non- const findShortestPaths()//
+/////////////////////////////////////////////////////////////////////////
+
+PathMap GridMap::findShortestPaths(ID start) {
+	return PathMap(*this, getShortestPaths(start, INT_MAX, MobilitySet(MobilityType::values::GROUNDED), false));
+}
+
+PathMap GridMap::findShortestPaths(size_t x, size_t y) {
+	return PathMap(*this, getShortestPaths((*this)[x][y]._id, INT_MAX, MobilitySet(MobilityType::values::GROUNDED), false));
+}
+
+PathMap GridMap::findShortestPaths(ID start, int max_move, MobilitySet mobility) {
+	bool intangible = mobility[MobilityType::values::PROJECTILE] || mobility[MobilityType::values::ETHEREAL];
+	return PathMap(*this, getShortestPaths(start, max_move, mobility, intangible));
+}
+
+PathMap GridMap::findShortestPaths(ID start, int max_move, MobilitySet mobility, bool intangible) {
+	return PathMap(*this, getShortestPaths(start, max_move, mobility, intangible));
+}
+
+PathMap GridMap::findShortestPaths(const Unit& unit) {
+	return PathMap(*this, getShortestPaths(unit));
+}
+
+//const findShortestPaths()//
+/////////////////////////////////////////////////////////////////////////
+constPathMap GridMap::findShortestPaths(ID start) const {
+	return constPathMap(*this, getShortestPaths(start, INT_MAX, MobilitySet(MobilityType::values::GROUNDED), false));
+}
+
+constPathMap GridMap::findShortestPaths(size_t x, size_t y) const {
+	return constPathMap(*this, getShortestPaths((*this)[x][y]._id, INT_MAX, MobilitySet(MobilityType::values::GROUNDED), false));
+}
+
+constPathMap GridMap::findShortestPaths(ID start, int max_move, MobilitySet mobility) const {
+	bool intangible = mobility[MobilityType::values::PROJECTILE] || mobility[MobilityType::values::ETHEREAL];
+	return constPathMap(*this, getShortestPaths(start, max_move, mobility, intangible));
+}
+
+constPathMap GridMap::findShortestPaths(ID start, int max_move, MobilitySet mobility, bool intangible) const {
+	return constPathMap(*this, getShortestPaths(start, max_move, mobility, intangible));
+}
+constPathMap GridMap::findShortestPaths(const Unit& unit) const {
+	return constPathMap(*this, getShortestPaths(unit));
+}
+
+// private helper methods for findShortestPaths() //
+/////////////////////////////////////////////////////////////////////////
+
+
+id_cost_map GridMap::getShortestPaths(const Unit& unit) const {
```

Comment:
Do we want to try to refactor Unit/Mobility so that this function can call the other?